### ISSUE TITLE
Rule proposal: `no-typeof-undefined`

### ISSUE THREAD

**Issue by B2o5T** [sindresorhus/eslint-plugin-unicorn#1492]
Less code written

## Fail

```js
typeof foo === 'undefined'
```

## Pass

```js
foo === undefined
```

**Comment by fisker** [comment]
How about `consistent-undefined-check`, allow one of these two? Personally I always use `typeof`.

**Comment by fisker** [comment]
When it's undefined variable, `foo === undefiend` will throw 

```text
foo === undefined
VM227:1 Uncaught ReferenceError: foo is not defined
    at <anonymous>:1:1
```

**Comment by B2o5T** [comment]
> When it's undefined variable, `foo === undefiend` will throw
> 
> ```
> foo === undefined
> VM227:1 Uncaught ReferenceError: foo is not defined
>     at <anonymous>:1:1
> ```

in node REPL, yes, but in real projects there is always enabled `no-undef` ESLint rule from `eslint:recommended` config, so this case will be reported

**Comment by fisker** [comment]
I don't think `no-undef` is safe, for example, when you write js for brower, the following won't report, because `AbortController` is considered "defined" as globals.

```js
if (AbortController !== undefined) {}
```

But when you run this code in old browsers, it will throw.

```js
if (typeof AbortController !== 'undefined') {}
```

This won't.

Of cause, `'AbortController' in globalThis` or `globalThis.AbortController` are safer. But nothing wrong to use `typeof`.

**Comment by B2o5T** [comment]
Ok, but we can check if the variable is declared in scope with `var`, `let`, or `const` keywords, if so, inform the suggestion, otherwise do nothing

**Comment by sindresorhus** [comment]
This is now accepted.

**Comment by sindresorhus** [comment]
> How about consistent-undefined-check, allow one of these two? Personally I always use typeof.

You are potentially silently hiding bugs by always using `typeof` though. I think it's generally better to only use it for things that are known to not exist, like when polyfilling. Any other times, IMHO, not using `typeof` is better.

**Comment by fisker** [comment]
I'm going to work on this. And I'm going to add an option to ignore global variables.

**Comment by fisker** [comment]
I've sent #1966, we'll check all `typeof ANYTHING === 'undefined'` by default.

When set `{checkGlobalVariables: false}`, the rule will ignore variables if it's not defined in file, including unresolved and known globals. `typeof foo === 'undefined'` and `typeof Array === 'undefined'` are both valid.

WDYT? @zloirock @papb

Or do you still prefer use `typeof`?

**Comment by zloirock** [comment]
@fisker I think that `checkGlobalVariables: false` should be set by default or this rule should be missed in the recommended.

**Comment by fisker** [comment]
I don't really mind, I'll use `checkGlobalVariables: false` anyway. @sindresorhus your call.

**Comment by sindresorhus** [comment]
I'm fine with having it `false` by default.

**Comment by fisker** [comment]
Changed `checkGlobalVariables` to `false` by default, also use suggestions for global variables to be safe.
